dataframe-csv/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/io/readDelim.kt

```diff
@@ -0,0 +1,343 @@
+import org.jetbrains.kotlinx.dataframe.io.skippingBomCharacters
+import org.jetbrains.kotlinx.dataframe.io.toKType
+import org.jetbrains.kotlinx.dataframe.io.useDecompressed
+import java.io.InputStream
+import java.math.BigDecimal
+import java.net.URL
+import kotlin.reflect.KType
+import kotlin.reflect.full.withNullability
+import kotlin.reflect.typeOf
+import kotlin.time.Duration
+
+/**
+ * Implementation to read delimiter-separated data from an [InputStream] based on the Deephaven CSV library.
+ *
+ * @include [INPUT_STREAM_READ]
+ * @param delimiter The field delimiter character. The default is ',' for CSV, '\t' for TSV.
+ * @include [HEADER]
+ * @include [COMPRESSION]
+ * @include [COL_TYPES]
+ * @include [SKIP_LINES]
+ * @include [READ_LINES]
+ * @include [HAS_FIXED_WIDTH_COLUMNS]
+ * @include [FIXED_COLUMN_WIDTHS]
+ * @include [PARSER_OPTIONS]
+ * @include [IGNORE_EMPTY_LINES]
+ * @include [ALLOW_MISSING_COLUMNS]
+ * @include [IGNORE_EXCESS_COLUMNS]
+ * @include [QUOTE]
+ * @include [IGNORE_SURROUNDING_SPACES]
+ * @include [TRIM_INSIDE_QUOTED]
+ * @include [PARSE_PARALLEL]
+ * @include [ADJUST_CSV_SPECS]
+ */
+internal fun readDelimImpl(
+    inputStream: InputStream,
+    delimiter: Char,
+    header: List<String>,
+    hasFixedWidthColumns: Boolean,
+    fixedColumnWidths: List<Int>,
+    compression: Compression<*>,
+    colTypes: Map<String, ColType>,
+    skipLines: Long,
+    readLines: Long?,
+    parserOptions: ParserOptions,
+    ignoreEmptyLines: Boolean,
+    allowMissingColumns: Boolean,
+    ignoreExcessColumns: Boolean,
+    quote: Char,
+    ignoreSurroundingSpaces: Boolean,
+    trimInsideQuoted: Boolean,
+    parseParallel: Boolean,
+    adjustCsvSpecs: AdjustCsvSpecs,
+): DataFrame<*> {
+    // set up the csv specs
+    val csvSpecs = with(CsvSpecs.builder()) {
+        customDoubleParser(DataFrameCustomDoubleParser(parserOptions))
+        nullValueLiterals(parserOptions.nullStrings ?: DEFAULT_NULL_STRINGS)
+        headerLegalizer(::legalizeHeader)
+        numRows(readLines ?: Long.MAX_VALUE)
+        ignoreEmptyLines(ignoreEmptyLines)
+        allowMissingColumns(allowMissingColumns)
+        ignoreExcessColumns(ignoreExcessColumns)
+        if (!hasFixedWidthColumns) delimiter(delimiter)
+        quote(quote)
+        ignoreSurroundingSpaces(ignoreSurroundingSpaces)
+        trim(trimInsideQuoted)
+        concurrent(parseParallel)
+        header(header)
+        hasFixedWidthColumns(hasFixedWidthColumns)
+        if (hasFixedWidthColumns && fixedColumnWidths.isNotEmpty()) fixedColumnWidths(fixedColumnWidths)
+        skipLines(takeHeaderFromCsv = header.isEmpty(), skipLines = skipLines)
+
+        // Deephaven's LocalDateTime parser is unconfigurable, so if the user provides a locale, pattern, or formatter,
+        // we must use our own parser for LocalDateTime and let Deephaven read them as Strings.
+        val useDeepHavenLocalDateTime = with(parserOptions) {
```

Comment:
Probably yes, however, in theory the user should not be able to tell the difference, only in speed.

- If they provide no configuration, it will use the fast LocalDateTime parser of deephaven and if that fails, it will remain a String and our slower LocalDateTime parser will try it too.

- If they do provide configuration, then deephaven's parser cannot be used and our slower LocalDateTime parser will be used.

In both cases, the user will get the expected result and where possible we save some performance. But we cannot "recommend" to not specify a configuration to gain performance, because that's dependent on the data the user has to read. All we can do is mention it somewhere in the documentation.